push/push.go

```diff
@@ -0,0 +1,80 @@
+package push
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/provider"
+	"github.com/evcc-io/evcc/util"
+)
+
+func init() {
+	registry.AddCtx(api.Custom, NewConfigurableFromConfig)
+}
+
+// NewConfigurableFromConfig creates Messenger from config
+func NewConfigurableFromConfig(ctx context.Context, other map[string]interface{}) (Messenger, error) {
+	var cc struct {
+		Send     provider.Config
+		Encoding string
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	send, err := provider.NewStringSetterFromConfig(ctx, "send", cc.Send)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewConfigurable(send, cc.Encoding)
+}
+
+// NewConfigurable creates a new Messenger
+func NewConfigurable(send func(string) error, encoding string) (*Push, error) {
+	m := &Push{
+		log:      util.NewLogger("push"),
+		send:     send,
+		encoding: strings.ToLower(encoding),
+	}
+	return m, nil
+}
+
+// Push is a configurable Messenger implementation
+type Push struct {
+	log      *util.Logger
+	send     func(string) error
+	encoding string
+}
+
+// Send implements the Messenger interface
+func (m *Push) Send(title, msg string) {
+	var res string
+
+	switch m.encoding {
+	case "json":
+		b, _ := json.Marshal(struct {
+			Title string `json:"title"`
+			Msg   string `json:"msg"`
+		}{
+			Title: title,
+			Msg:   msg,
+		})
+		res = string(b)
+	case "csv":
+		res = title + "," + msg
```

Comment:
This is too simplistic, especially when `title` and `message` can contain commas (to tabs for `tsv`). Do we really need csv and tsv ? I can't see any good use case that can not be solved with a stronger typed representation (like json).

Reply:
Does it hurt? Since users do control the message format this can be handled by crafting the right kind of message. Happy to improve this by doing a better encoding.

Reply:
Nvm. We can just use the cvswriter for this as well.

Reply:
> Does it hurt? Since users do control the message format this can be handled by crafting the right kind of message. Happy to improve this by doing a better encoding.

Well, I've been in the open-source software business for too long, I guess. I've learned over the long run to respect YAGNI and better leave out things than anticipate anything the people _might_ use at some point in time. It's easy to add these features later anyway if needed, but if nobody uses them, you will carry that baggage with you forever. Ideally, you add some tests for it (especially for corner cases and strange escape rules), and you need some documentation. I became a fanboy of _less is more_ most of the time.

Or let me ask the other way around: Which use case do you think needs CSV, TSV, or title—only for sending out to an endpoint? Couldn't you just do it all yourself when defining the `msg`, even with the proper templating support?

In the end, it's your call anyway, as I'm like many other contributors: Here for some time to help, but likely not forever. Just sharing some stuff.

Reply:
> Or let me ask the other way around: Which use case do you think needs CSV, TSV, or title—only for sending out to an endpoint?

Anybody using the `script` messenger right now.

Reply:
I don't think so. 

As far as I see, does the script provider *not* implement `StringSetter` at all (so it wouldn't work anyway as is), and if StringSetter would be implemented in `provider/script.go`, you probably would replace a single argument in `${param}` in the script definition, so having an unsafe trivial encoding via un-escaped CSV or TSV doesn't provide any value over just sending the plain message and ignoring the title (especially when you have to parse the input in a shell script anyway). This is not user-friendly, so I would keep it simplistic and ignore the title (or maybe even throw an error if a title is defined using a custom provider).

Also, what do you do when the title is empty? This will lead to strange CSV and TSV representations (like a leading `,` or leading tab). But you should allow for empty titles for this kind of integration, as this would be the most natural way to define no `title` and a JSON `msg` when using a custom provider (you can always add a `title:` in that `msg` that the user defines if you like)

Have you ever integration-tested this PR with a script (or any other) provider? I will try it out now with mqtt and will report back ...

Reply:
Ah, you probably thought about using 

```
cmdline: script ${send}
```

and then get title and message as two separate arguments when using `tsv`.  But what are you doing then `msg` contains spaces (which is very likely) ?

I would guess that most of the time, one would use `cmdline: script "${send}"` to get the whole input as a single argument to the script. CSV and TSV are common but complicate things in that case. As said, I would always leave out the title for custom providers.